### ISSUE TITLE
ApplicationLauncherImpl: Illegal reflective access

### ISSUE THREAD

**Issue by Zhiyuan-Amos** [TestFX/TestFX#553]
Hello, I'm using Gradle version 4.6 and testfx-core version 4.0.12-alpha, and I received an `illegal reflective access operation` warning message as seen below:

`WARNING: Illegal reflective access by org.testfx.toolkit.impl.ApplicationLauncherImpl (file:/C:/Users/appveyor/.gradle/caches/modules-2/files-2.1/org.testfx/testfx-core/4.0.12-alpha/55d68835a7f5c0f3b5630a2a542e0880f2548e7f/testfx-core-4.0.12-alpha.jar) to field com.sun.glass.ui.PlatformFactory.instance`

Making a quick guess, this method seems to be the one causing the issue:
https://github.com/TestFX/TestFX/blob/2bd619f1c1ecda95a934513cd54fc3064197e537/subprojects/testfx-core/src/main/java/org/testfx/toolkit/impl/ApplicationLauncherImpl.java#L63-L68

**Comment by brcolow** [comment]
What version of the JDK?

**Comment by brcolow** [comment]
Yes, that warning is valid. But thankfully, just a warning (is it still a warning on Java 11...I don't know). I am not sure how it can be fixed given the current `com.sun.glass.ui.PlatformFactory` code. I believe it will require a change to upstream OpenJFX.

**Comment by kevinrushforth** [comment]
Doesn't System.setProperty("glass.platform", "Monocle") already give you what you need?

**Comment by brcolow** [comment]
I don't believe it does. Taking away the reflection we use (which my suggested changes would get rid of).

```java
    private void assignMonoclePlatform() throws Exception {
        Class<?> platformFactoryClass = Class.forName("com.sun.glass.ui.PlatformFactory");
        Object platformFactoryImpl = Class.forName("com.sun.glass.ui.monocle.MonoclePlatformFactory")
                .getDeclaredConstructor().newInstance();
        assignPrivateStaticField(platformFactoryClass, "instance", platformFactoryImpl);
    }
```

but with the glass.platform=Monocle property, I get the "Failed to load Glass factory class" error:

https://travis-ci.org/TestFX/TestFX/jobs/363592359

and the headless builds fail.

It would definitely make sense to re-use that property instead of the "com.sun.javafx.headless" one I made up, but I don't think it is enough currently to force load Monocle on any platform (macOS, Windows, *nix).

**Comment by brcolow** [comment]
I am currently in the process of debugging why this is the case.

**Comment by kevinrushforth** [comment]
I'll take a look at it (hopefully tomorrow). It does sound like a ClassLoader issue, but I need some time to dive into it. A quick question: are you putting the Monocle classes on the application classpath / module path? Or are you patching the javafx modules and including them into the javafx.graphics module?

**Comment by brcolow** [comment]
I believe I have tried both ways (the Monocle classes are on the classpath because we have a gradle dependency on them), and I tried patching with `--patch-module` but I may have missed other necessary arguments (`--add-reads`? `--add-modules`? I don't know....).

**Comment by kevinrushforth** [comment]
The problem (actually a couple of problems) is in this code:

```
        String platform = Platform.determinePlatform();
        String factory = "com.sun.glass.ui." + platform.toLowerCase(Locale.ROOT) + "."+ platform + "PlatformFactory";
        // System.out.println("Loading Glass Factory " + factory);
--->    Class c = Class.forName(factory);
        instance = (PlatformFactory) c.newInstance();
```

The call to Class.forName uses the class loader of the enclosing Class (PlatformFactory), which won't find any class on the classpath. It still ought to work with --patch-module, since that's how we run using Monocle-based system tests using 'gradle -PFULL_TEST=true test'.

In any case, the code should be fixed in one (or maybe both) of the following two ways:

1. Use the correct class loading pattern (e.g., as used in Toolkit.java or Skin.java) :

```
        ClassLoader loader = PlatformFactory.class.get
        Class c = Class.forName(factory, false, loader);
        instance = (PlatformFactory) c.newInstance();

        try {
            // Try our class loader first
            final ClassLoader loader = PlatformFactory.class.getClassLoader();
            clz = Class.forName(factory, false, loader);
        } catch (ClassNotFoundException e) {
            // fall back and try the application class loader
            final ClassLoader loader = Thread.currentThread().getContextClassLoader();
            clz = Class.forName(factory, false, loader);
        }
```

As an aside, we should also check that PlatformFactory is assignable from the returned class before calling newInstance.


2. Ship Monocle with FX as part of the runtime even on desktop platforms

This would only be used for headless testing on desktop platforms, so might be overkill. Still, there might be some advantages.

**Comment by brcolow** [comment]
Thanks so much for digging into this. Do you think you will have time to open a JBS bug and work on it or is it something you would like me to open a PR for on javafxports/openjdk-jfx GitHub. I'm happy to help, if necessary.

**Comment by kevinrushforth** [comment]
Your use of `--patch-module` looks correct, so I don't know why it doesn't work. You can take a look at build/testrun.args (which is generated by running 'gradle test') to see how we patch the modules for testing.

I can open a JBS bug for this, but I won't have time to work on it until after the separation of FX from the JDK. One thing to note is that this bug might actually turn out not to matter once FX is loaded by the App ClassLoader (will need to test).

**Comment by kevinrushforth** [comment]
I just did a quick test, and I am able to load Monocle by using `--patch-module` either by pointing to a jar or an exploded directory containing the class files. Note that in either case, the root must be the same one you would use for a classpath. That is, the root should contain the "com" directory and not contain "javafx.graphics".
